### PR TITLE
Fix migration state problems on node shutdown and join

### LINKED ISSUE #6655: [TEST-FAILURE] MigrationAwareServiceTest.testPartitionDataSize_whenNodeGracefullyShutdown
```
java.lang.AssertionError: expected:<271> but was:<215>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at org.junit.Assert.assertEquals(Assert.java:631)
	at com.hazelcast.spi.MigrationAwareServiceTest$3.run(MigrationAwareServiceTest.java:222)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:825)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:839)
	at com.hazelcast.spi.MigrationAwareServiceTest.assertSize(MigrationAwareServiceTest.java:211)
	at com.hazelcast.spi.MigrationAwareServiceTest.testPartitionDataSize_whenNodeGracefullyShutdown(MigrationAwareServiceTest.java:160)
```

https://hazelcast-l337.ci.cloudbees.com/job/Hazelcast-3.x-IbmJDK1.7/com.hazelcast$hazelcast/756/testReport/junit/com.hazelcast.spi/MigrationAwareServiceTest/testPartitionDataSize_whenNodeGracefullyShutdown/

### LINKED ISSUE #6656: BackupTest.testGracefulShutdown3
```
java.lang.AssertionError: expected:<50000> but was:<49269>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at org.junit.Assert.assertEquals(Assert.java:631)
	at com.hazelcast.map.BackupTest$1.run(BackupTest.java:71)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:825)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:839)
	at com.hazelcast.map.BackupTest.checkSizeEventually(BackupTest.java:68)
	at com.hazelcast.map.BackupTest.testGracefulShutdown3(BackupTest.java:230)
```

https://hazelcast-l337.ci.cloudbees.com/job/Hazelcast-3.x-nightly/com.hazelcast$hazelcast/762/testReport/junit/com.hazelcast.map/BackupTest/testGracefulShutdown3/

### LINKED ISSUE #6663: A bug in nodestate transition logic?
this code has an ability to freeze my member in the latest master (a7788ba319) :

```java
public class MapTest {
    public static void main(String[] args) throws InterruptedException {
        HazelcastInstance i1 = Hazelcast.newHazelcastInstance();
        HazelcastInstance i2 = Hazelcast.newHazelcastInstance();

        final IMap<Integer, Integer> map = i1.getMap("map");

        int itemsCount = 1000000;
        for (int i = 0; i < itemsCount; i++) {
            map.put(i, i);
            if (i % 10000 == 0) {
                System.out.println("Loading progress: " + i + " / " + itemsCount);
            }
        }


        int stressThreadsCount = 5;
        for (int i = 0; i < stressThreadsCount; i++) {
            new StressThread(map).start();
        }

        for (;;) {
            i2.shutdown();
            i2 = Hazelcast.newHazelcastInstance();
        }
    }

    private static class StressThread extends Thread {
        private final IMap<Integer, Integer> map;

        public StressThread(IMap<Integer, Integer> map) {
            this.map = map;
        }

        @Override
        public void run() {
            for (;;) {
                map.containsKey(0);
            }
        }
    }
}
```

I can see this in the log:
```
WARNING: [10.8.0.10]:5702 [dev] [3.6-RC1-SNAPSHOT] Partition table received before startup is completed. Caller: Address[10.8.0.10]:5701
Nov 05, 2015 7:36:55 PM com.hazelcast.partition.InternalPartitionService
INFO: [10.8.0.10]:5702 [dev] [3.6-RC1-SNAPSHOT] Waiting for active migration tasks: 42
Nov 05, 2015 7:36:56 PM com.hazelcast.partition.InternalPartitionService
INFO: [10.8.0.10]:5702 [dev] [3.6-RC1-SNAPSHOT] Waiting for active migration tasks: 42
Nov 05, 2015 7:36:57 PM com.hazelcast.partition.InternalPartitionService
INFO: [10.8.0.10]:5702 [dev] [3.6-RC1-SNAPSHOT] Waiting for active migration tasks: 42
Nov 05, 2015 7:36:58 PM com.hazelcast.partition.InternalPartitionService
INFO: [10.8.0.10]:5702 [dev] [3.6-RC1-SNAPSHOT] Waiting for active migration tasks: 42
Nov 05, 2015 7:36:59 PM com.hazelcast.partition.InternalPartitionService
INFO: [10.8.0.10]:5702 [dev] [3.6-RC1-SNAPSHOT] Waiting for active migration tasks: 42
Nov 05, 2015 7:37:00 PM com.hazelcast.partition.InternalPartitionService
INFO: [10.8.0.10]:5702 [dev] [3.6-RC1-SNAPSHOT] Waiting for active migration tasks: 42
Nov 05, 2015 7:37:01 PM com.hazelcast.partition.InternalPartitionService
INFO: [10.8.0.10]:5702 [dev] [3.6-RC1-SNAPSHOT] Waiting for active migration tasks: 42
Nov 05, 2015 7:37:02 PM com.hazelcast.partition.InternalPartitionService
INFO: [10.8.0.10]:5702 [dev] [3.6-RC1-SNAPSHOT] Waiting for active migration tasks: 42
Nov 05, 2015 7:37:03 PM com.hazelcast.partition.InternalPartitionService
INFO: [10.8.0.10]:5702 [dev] [3.6-RC1-SNAPSHOT] Waiting for active migration tasks: 42
Nov 05, 2015 7:37:04 PM com.hazelcast.partition.InternalPartitionService
INFO: [10.8.0.10]:5702 [dev] [3.6-RC1-SNAPSHOT] Waiting for active migration tasks: 42
Nov 05, 2015 7:37:05 PM com.hazelcast.partition.InternalPartitionService
INFO: [10.8.0.10]:5702 [dev] [3.6-RC1-SNAPSHOT] Waiting for active migration tasks: 42
Nov 05, 2015 7:37:06 PM com.hazelcast.partition.InternalPartitionService
```
it's repeating over and over again, but it seems it never recovers.

### PR DESCRIPTION
**Pull request by metanet**
* Do not trigger RepartitioningTask if cluster state is not active. It causes infinite retries and prevents nodes to shutdown since they think that there is an ongoing migration
* Introduce NodeExtension.isStartCompleted() method and use it in PartitionStateOperation instead of checking NodeState. It is a no-op method for OSS and has an implementation on enterprise.
* Send initial partition table with finalize-join operation 

Fixes #6663
Fixes #6655
Fixes #6656

### CODE REVIEW

File: hazelcast/src/test/java/com/hazelcast/partition/impl/FrozenPartitionTableTest.java [hunk 1]
```diff
@@ -40,6 +42,7 @@ public void partitionTable_isFrozen_whenNodesLeave_duringClusterStateIsPassive()
         testPartitionTableIsFrozenDuring(ClusterState.PASSIVE);
     }
 
+    @Repeat(100)
```
Comment (mdogan) [inline hazelcast/src/test/java/com/hazelcast/partition/impl/FrozenPartitionTableTest.java]: leftover

**Comment by mdogan** [comment]
:+1:

File: hazelcast/src/test/java/com/hazelcast/partition/impl/FrozenPartitionTableTest.java (hunk 1 above)
Comment (mdogan) [inline hazelcast/src/test/java/com/hazelcast/partition/impl/FrozenPartitionTableTest.java]: I can remove _repeat_ annotation with a manual merge.

**Comment by gurbuzali** [comment]
Looks good to be merged

**Pull request merged**